File: KaiFlow.Timesheets.Maui/Models/PunchSession.cs
namespace KaiFlow.Timesheets.Models;

/// <summary>
/// Aggregates a clock-in + clock-out pair into a payable session.
/// Computed client-side from pairs of TimePunch records.
/// </summary>
public class PunchSession
{
    public Guid EmployeeId { get; init; }
    public string EmployeeName { get; init; } = "";
    public string? EmployeeCode { get; init; }
    public DateTime ClockIn { get; init; }
    public DateTime? ClockOut { get; init; }
    public double DailyHours { get; init; } = 8.0;
    public double OvertimeRate { get; init; } = 1.5;
    public double HourlyRate { get; init; }
    public Guid? JobId { get; init; }
    public string? Notes { get; init; }

    /// <summary>Synthetic row for an HR-marked or self-reported absence (no punches).</summary>
    public bool IsAbsentDay { get; init; }
    /// <summary>Synthetic row for an approved leave day (no punches).</summary>
    public bool IsLeaveDay { get; init; }
    public string? StatusNote { get; init; }

    public bool IsNonWorkDay => IsAbsentDay || IsLeaveDay;

    // Shift template & company attendance rules (null = fall back to DailyHours)
    public EmployeeShiftTemplate? ShiftTemplate { get; init; }
    public int LateThresholdMinutes { get; init; } = 30;
    public int OtStartAfterMinutes { get; init; } = 30;

    // Clock-in location
    public double? ClockInLatitude { get; init; }
    public double? ClockInLongitude { get; init; }
    public string? ClockInAddress { get; init; }

    // Clock-out location
    public double? ClockOutLatitude { get; init; }
    public double? ClockOutLongitude { get; init; }
    public string? ClockOutAddress { get; init; }

    // ── Raw elapsed ─────────────────────────────────────────────────────────

    public TimeSpan Duration => ClockOut.HasValue
        ? ClockOut.Value - ClockIn
        : DateTime.UtcNow - ClockIn;

    public double TotalHours => Duration.TotalHours;

    // ── Late detection ──────────────────────────────────────────────────────

    public bool IsLate
    {
        get
        {
            if (IsNonWorkDay) return false;
            if (ShiftTemplate == null) return false;
            var localIn = TimeOnly.FromDateTime(ClockIn.ToLocalTime());
            var minsLate = (localIn - ShiftTemplate.StartTime).TotalMinutes;
            return minsLate > LateThresholdMinutes;
        }
    }

    public int LateMinutes
    {
        get
        {
            if (!IsLate || ShiftTemplate == null) return 0;
            var localIn = TimeOnly.FromDateTime(ClockIn.ToLocalTime());
            return (int)(localIn - ShiftTemplate.StartTime).TotalMinutes;
        }
    }

    public string LateFlag
    {
        get
        {
            if (!IsLate) return "";
            var h = LateMinutes / 60;
            var m = LateMinutes % 60;
            var parts = h > 0 ? (m > 0 ? $"{h}h {m}m" : $"{h}h") : $"{m}m";
            return $"Late {parts}";
        }
    }

    // ── Early departure ──────────────────────────────────────────────────────

    public bool IsLeftEarly
    {
        get
        {
            if (ShiftTemplate == null || !ClockOut.HasValue) return false;
            var localOut = ClockOut.Value.ToLocalTime();
            var shiftEnd = localOut.Date.Add(ShiftTemplate.EndTime.ToTimeSpan());
            if (shiftEnd < BillingStart) shiftEnd = shiftEnd.AddDays(1);
            return localOut < shiftEnd;
        }
    }

    public int EarlyMinutes
    {
        get
        {
            if (!IsLeftEarly || ShiftTemplate == null || !ClockOut.HasValue) return 0;
            var localOut = ClockOut.Value.ToLocalTime();
            var shiftEnd = localOut.Date.Add(ShiftTemplate.EndTime.ToTimeSpan());
            if (shiftEnd < BillingStart) shiftEnd = shiftEnd.AddDays(1);
            return (int)(shiftEnd - localOut).TotalMinutes;
        }
    }

    public string EarlyFlag
    {
        get
        {
            if (!IsLeftEarly) return "";
            var h = EarlyMinutes / 60;
            var m = EarlyMinutes % 60;
            var parts = h > 0 ? (m > 0 ? $"{h}h {m}m" : $"{h}h") : $"{m}m";
            return $"Left Early {parts}";
        }
    }

    // ── Template-aware billing ───────────────────────────────────────────────

    // Where billing starts: shift-start if on time/early, actual clock-in if late
    private DateTime BillingStart
    {
        get
        {
            if (ShiftTemplate == null) return ClockIn;
            var localIn = ClockIn.ToLocalTime();
            return IsLate
                ? localIn
                : localIn.Date.Add(ShiftTemplate.StartTime.ToTimeSpan());
        }
    }

    public double RegularHours
    {
        get
        {
            if (IsNonWorkDay) return 0;

            if (ShiftTemplate == null)
                return Math.Min(TotalHours, DailyHours);

            if (!ClockOut.HasValue)
                return Math.Max(0, (DateTime.Now - BillingStart).TotalHours
                    - ShiftTemplate.TotalBreakMinutes / 60.0);

            var localOut = ClockOut.Value.ToLocalTime();
            var paidElapsed = (localOut - BillingStart).TotalHours
                - ShiftTemplate.TotalBreakMinutes / 60.0;
            return Math.Max(0, Math.Min(paidElapsed, ShiftTemplate.PaidHours));
        }
    }

    public double OvertimeHours
    {
        get
        {
            if (IsNonWorkDay) return 0;

            if (ShiftTemplate == null)
                return Math.Max(0, TotalHours - DailyHours);

            if (!ClockOut.HasValue) return 0;

            var localOut = ClockOut.Value.ToLocalTime();
            var shiftEnd = localOut.Date.Add(ShiftTemplate.EndTime.ToTimeSpan());
            if (shiftEnd < BillingStart) shiftEnd = shiftEnd.AddDays(1);

            var minutesPastEnd = (localOut - shiftEnd).TotalMinutes;
            return minutesPastEnd > OtStartAfterMinutes
                ? (minutesPastEnd - OtStartAfterMinutes) / 60.0
                : 0;
        }
    }

    public double RegularPay  => RegularHours * HourlyRate;
    public double OvertimePay => OvertimeHours * HourlyRate * OvertimeRate;
    public double TotalPay    => RegularPay + OvertimePay;

    public bool IsOpen      => !ClockOut.HasValue;
    public bool HasLocation => ClockInLatitude.HasValue && ClockInLongitude.HasValue;
    public string? MapsUrl  => HasLocation ? $"https://maps.google.com/?q={ClockInLatitude},{ClockInLongitude}" : null;

    public DateOnly Date => DateOnly.FromDateTime(ClockIn.ToLocalTime());

    // Display helpers
    public string TimeInDisplay => IsAbsentDay ? "Absent" : IsLeaveDay ? "On Leave" : ClockIn.ToLocalTime().ToString("h:mm tt");
    public string TimeOutDisplay => IsNonWorkDay ? "—" : ClockOut.HasValue ? ClockOut.Value.ToLocalTime().ToString("h:mm tt") : "—";
    public string DateDisplay     => ClockIn.ToLocalTime().ToString("ddd d MMM yyyy");
    public string TotalHrsDisplay => IsNonWorkDay ? "0.0h" : $"{(RegularHours + OvertimeHours):F1}h";
    public string TimeInColorHex  => IsAbsentDay ? "#EF4444" : IsLeaveDay ? "#F59E0B" : IsLate ? "#EF4444" : "#22C55E";
    public string InLocationDisplay  => IsNonWorkDay ? "—" : ClockInAddress ?? (ClockInLatitude.HasValue  ? $"{ClockInLatitude:F4}, {ClockInLongitude:F4}"  : "—");
    public string OutLocationDisplay => IsNonWorkDay ? "—" : ClockOutAddress ?? (ClockOutLatitude.HasValue ? $"{ClockOutLatitude:F4}, {ClockOutLongitude:F4}" : "—");
    public string DisplayNotes => IsNonWorkDay ? (StatusNote ?? (IsAbsentDay ? "Absent" : "On leave")) : (Notes ?? "—");

    // ── Build sessions from punch records ────────────────────────────────────

    public static List<PunchSession> Build(
        IEnumerable<TimePunch> punches,
        Dictionary<Guid, Employee>? employeeMap = null,
        Dictionary<Guid, EmployeeShiftTemplate>? templateMap = null,
        int lateThresholdMinutes = 30,
        int otStartAfterMinutes = 30)
    {
        var sessions = new List<PunchSession>();
        var grouped = punches.GroupBy(p => p.EmployeeId);

        foreach (var group in grouped)
        {
            Employee? emp = null;
            employeeMap?.TryGetValue(group.Key, out emp);

            EmployeeShiftTemplate? template = null;
            if (templateMap != null && emp?.ShiftTemplateId.HasValue == true)
                templateMap.TryGetValue(emp.ShiftTemplateId!.Value, out template);

            var sorted = group.OrderBy(p => p.DateTime).ToList();
            TimePunch? clockIn = null;

            foreach (var punch in sorted)
            {
                if (punch.PunchType == PunchType.In)
                {
                    if (clockIn != null)
                        sessions.Add(MakeSession(clockIn, null, group.Key, emp, template, lateThresholdMinutes, otStartAfterMinutes));
                    clockIn = punch;
                }
                else if (clockIn != null)
                {
                    sessions.Add(MakeSession(clockIn, punch, group.Key, emp, template, lateThresholdMinutes, otStartAfterMinutes));
                    clockIn = null;
                }
                else
                {
                    sessions.Add(MakeSession(null, punch, group.Key, emp, template, lateThresholdMinutes, otStartAfterMinutes));
                }
            }

            if (clockIn != null)
                sessions.Add(MakeSession(clockIn, null, group.Key, emp, template, lateThresholdMinutes, otStartAfterMinutes));
        }

        return sessions.OrderByDescending(s => s.ClockIn).ToList();
    }

    private static PunchSession MakeSession(
        TimePunch? clockIn, TimePunch? clockOut,
        Guid employeeId, Employee? emp,
        EmployeeShiftTemplate? template,
        int lateThreshold, int otStartAfter)
    {
        var outPunch = clockOut;
        return new PunchSession
        {
            EmployeeId        = employeeId,
            EmployeeName      = emp?.FullName ?? employeeId.ToString(),
            EmployeeCode      = emp?.EmployeeCode,
            ClockIn           = (clockIn?.DateTime ?? outPunch!.DateTime).ToUniversalTime(),
            ClockOut          = outPunch?.DateTime.ToUniversalTime(),
            HourlyRate        = emp?.HourlyRate ?? 0,
            DailyHours        = emp?.DailyHours ?? 8.0,
            OvertimeRate      = emp?.OvertimeRate > 0 ? emp.OvertimeRate : 1.5,
            ShiftTemplate     = template,
            LateThresholdMinutes = lateThreshold,
            OtStartAfterMinutes  = otStartAfter,
            JobId             = clockIn?.JobId,
            Notes             = outPunch?.Notes ?? clockIn?.Notes,
            ClockInLatitude   = clockIn?.Latitude,
            ClockInLongitude  = clockIn?.Longitude,
            ClockInAddress    = clockIn?.Address,
            ClockOutLatitude  = outPunch?.Latitude,
            ClockOutLongitude = outPunch?.Longitude,
            ClockOutAddress   = outPunch?.Address,
        };
    }

    public static PunchSession ForAbsentDay(
        Employee emp,
        DateOnly date,
        DailyAbsence absence,
        EmployeeShiftTemplate? template = null,
        int lateThresholdMinutes = 30,
        int otStartAfterMinutes = 30)
    {
        var localMidnight = date.ToDateTime(TimeOnly.MinValue);
        var note = absence.ReasonLabel;
        if (!string.IsNullOrWhiteSpace(absence.Note))
            note += $" – {absence.Note}";

        return new PunchSession
        {
            EmployeeId           = emp.Id,
            EmployeeName         = emp.FullName,
            EmployeeCode         = emp.EmployeeCode,
            ClockIn              = localMidnight.ToUniversalTime(),
            ClockOut             = localMidnight.ToUniversalTime(),
            HourlyRate           = emp.HourlyRate,
            DailyHours           = emp.DailyHours,
            OvertimeRate         = emp.OvertimeRate > 0 ? emp.OvertimeRate : 1.5,
            ShiftTemplate        = template,
            LateThresholdMinutes = lateThresholdMinutes,
            OtStartAfterMinutes  = otStartAfterMinutes,
            IsAbsentDay          = true,
            StatusNote           = note,
            Notes                = note,
        };
    }

    public static PunchSession ForLeaveDay(
        Employee emp,
        DateOnly date,
        LeaveRequest leave,
        EmployeeShiftTemplate? template = null,
        int lateThresholdMinutes = 30,
        int otStartAfterMinutes = 30)
    {
        var localMidnight = date.ToDateTime(TimeOnly.MinValue);
        var note = leave.LeaveType;

        return new PunchSession
        {
            EmployeeId           = emp.Id,
            EmployeeName         = emp.FullName,
            EmployeeCode         = emp.EmployeeCode,
            ClockIn              = localMidnight.ToUniversalTime(),
            ClockOut             = localMidnight.ToUniversalTime(),
            HourlyRate           = emp.HourlyRate,
            DailyHours           = emp.DailyHours,
            OvertimeRate         = emp.OvertimeRate > 0 ? emp.OvertimeRate : 1.5,
            ShiftTemplate        = template,
            LateThresholdMinutes = lateThresholdMinutes,
            OtStartAfterMinutes  = otStartAfterMinutes,
            IsLeaveDay           = true,
            StatusNote           = note,
            Notes                = note,
        };
    }
}

File: kaisync-web/src/lib/punch-session.ts
/**
 * Client-side punch session builder — mirrors MAUI PunchSession.Build.
 * Pairs clock-in/out, computes regular/OT hours, late/early flags.
 */

export type PunchLike = {
  id?: string
  employee_id?: string
  type: 'in' | 'out' | string
  date_time: string
  latitude?: number | null
  longitude?: number | null
  address?: string | null
  job_id?: string | null
  notes?: string | null
}

export type ShiftTemplateLike = {
  id: string
  start_time?: string | null // "HH:mm:ss" or "HH:mm"
  end_time?: string | null
  break_minutes?: number | null
  total_break_minutes?: number | null
}

export type PunchSessionOptions = {
  employeeId: string
  employeeName?: string
  dailyHours?: number
  lateThresholdMinutes?: number
  otStartAfterMinutes?: number
  shiftTemplate?: ShiftTemplateLike | null
}

export type PunchSessionRow = {
  employeeId: string
  employeeName: string
  clockIn: Date
  clockOut: Date | null
  jobId: string | null
  notes: string | null
  clockInAddress: string | null
  clockOutAddress: string | null
  clockInLat: number | null
  clockInLng: number | null
  clockOutLat: number | null
  clockOutLng: number | null
  isAbsentDay: boolean
  isLeaveDay: boolean
  statusNote: string | null
  regularHours: number
  overtimeHours: number
  isLate: boolean
  lateMinutes: number
  isLeftEarly: boolean
  earlyMinutes: number
  isOpen: boolean
}

function parseTimeOnly(raw: string | null | undefined): { h: number; m: number } | null {
  if (!raw) return null
  const m = raw.match(/^(\d{1,2}):(\d{2})/)
  if (!m) return null
  return { h: Number(m[1]), m: Number(m[2]) }
}

function atLocalTime(day: Date, t: { h: number; m: number }): Date {
  const d = new Date(day)
  d.setHours(t.h, t.m, 0, 0)
  return d
}

function paidHoursOf(template: ShiftTemplateLike): number {
  const start = parseTimeOnly(template.start_time)
  const end = parseTimeOnly(template.end_time)
  if (!start || !end) return 8
  let span = (end.h * 60 + end.m) - (start.h * 60 + start.m)
  if (span <= 0) span += 24 * 60
  const brk = Number(template.total_break_minutes ?? template.break_minutes ?? 0)
  return Math.max(0, span / 60 - brk / 60)
}

function breakMinutesOf(template: ShiftTemplateLike | null | undefined): number {
  if (!template) return 0
  return Number(template.total_break_minutes ?? template.break_minutes ?? 0)
}

function computeMetrics(
  clockIn: Date,
  clockOut: Date | null,
  opts: PunchSessionOptions,
): Pick<PunchSessionRow, 'regularHours' | 'overtimeHours' | 'isLate' | 'lateMinutes' | 'isLeftEarly' | 'earlyMinutes'> {
  const dailyHours = opts.dailyHours ?? 8
  const lateThreshold = opts.lateThresholdMinutes ?? 30
  const otAfter = opts.otStartAfterMinutes ?? 30
  const template = opts.shiftTemplate ?? null
  const now = new Date()
  const totalHours = ((clockOut ?? now).getTime() - clockIn.getTime()) / 3600000

  let isLate = false
  let lateMinutes = 0
  let isLeftEarly = false
  let earlyMinutes = 0
  let regularHours = 0
  let overtimeHours = 0

  const startT = template ? parseTimeOnly(template.start_time) : null
  const endT = template ? parseTimeOnly(template.end_time) : null

  if (template && startT) {
    const shiftStart = atLocalTime(clockIn, startT)
    const minsLate = (clockIn.getTime() - shiftStart.getTime()) / 60000
    if (minsLate > lateThreshold) {
      isLate = true
      lateMinutes = Math.floor(minsLate)
    }
  }

  if (template && endT && clockOut) {
    let shiftEnd = atLocalTime(clockOut, endT)
    const billingStart = isLate || !startT
      ? clockIn
      : atLocalTime(clockIn, startT)
    if (shiftEnd.getTime() < billingStart.getTime()) {
      shiftEnd = new Date(shiftEnd.getTime() + 24 * 3600000)
    }
    if (clockOut.getTime() < shiftEnd.getTime()) {
      isLeftEarly = true
      earlyMinutes = Math.floor((shiftEnd.getTime() - clockOut.getTime()) / 60000)
    }
  }

  if (!template) {
    regularHours = Math.min(totalHours, dailyHours)
    overtimeHours = Math.max(0, totalHours - dailyHours)
  } else if (startT) {
    const billingStart = isLate ? clockIn : atLocalTime(clockIn, startT)
    const brk = breakMinutesOf(template) / 60
    const paidCap = paidHoursOf(template)
    if (!clockOut) {
      regularHours = Math.max(0, (now.getTime() - billingStart.getTime()) / 3600000 - brk)
    } else {
      const paidElapsed = (clockOut.getTime() - billingStart.getTime()) / 3600000 - brk
      regularHours = Math.max(0, Math.min(paidElapsed, paidCap))
      if (endT) {
        let shiftEnd = atLocalTime(clockOut, endT)
        if (shiftEnd.getTime() < billingStart.getTime()) {
          shiftEnd = new Date(shiftEnd.getTime() + 24 * 3600000)
        }
        const minutesPastEnd = (clockOut.getTime() - shiftEnd.getTime()) / 60000
        overtimeHours = minutesPastEnd > otAfter ? (minutesPastEnd - otAfter) / 60 : 0
      }
    }
  }

  return {
    regularHours: Math.round(regularHours * 100) / 100,
    overtimeHours: Math.round(overtimeHours * 100) / 100,
    isLate,
    lateMinutes,
    isLeftEarly,
    earlyMinutes,
  }
}

function makeSession(
  clockIn: PunchLike | null,
  clockOut: PunchLike | null,
  opts: PunchSessionOptions,
): PunchSessionRow {
  const inDt = new Date((clockIn ?? clockOut)!.date_time)
  const outDt = clockOut ? new Date(clockOut.date_time) : null
  const metrics = computeMetrics(inDt, outDt, opts)
  return {
    employeeId: opts.employeeId,
    employeeName: opts.employeeName ?? 'Employee',
    clockIn: inDt,
    clockOut: outDt,
    jobId: clockIn?.job_id ?? null,
    notes: clockOut?.notes ?? clockIn?.notes ?? null,
    clockInAddress: clockIn?.address ?? null,
    clockOutAddress: clockOut?.address ?? null,
    clockInLat: clockIn?.latitude ?? null,
    clockInLng: clockIn?.longitude ?? null,
    clockOutLat: clockOut?.latitude ?? null,
    clockOutLng: clockOut?.longitude ?? null,
    isAbsentDay: false,
    isLeaveDay: false,
    statusNote: null,
    ...metrics,
    isOpen: !outDt,
  }
}

/** Pair punches into sessions (MAUI PunchSession.Build). */
export function buildPunchSessions(
  punches: PunchLike[],
  opts: PunchSessionOptions,
): PunchSessionRow[] {
  const sorted = [...punches].sort(
    (a, b) => new Date(a.date_time).getTime() - new Date(b.date_time).getTime(),
  )
  const sessions: PunchSessionRow[] = []
  let clockIn: PunchLike | null = null

  for (const punch of sorted) {
    const type = (punch.type ?? '').toLowerCase()
    if (type === 'in') {
      if (clockIn) sessions.push(makeSession(clockIn, null, opts))
      clockIn = punch
    } else if (clockIn) {
      sessions.push(makeSession(clockIn, punch, opts))
      clockIn = null
    } else {
      sessions.push(makeSession(null, punch, opts))
    }
  }
  if (clockIn) sessions.push(makeSession(clockIn, null, opts))

  return sessions.sort((a, b) => b.clockIn.getTime() - a.clockIn.getTime())
}

export function absentDaySession(
  dateStr: string,
  reason: string,
  opts: PunchSessionOptions,
): PunchSessionRow {
  const d = new Date(`${dateStr}T00:00:00`)
  return {
    employeeId: opts.employeeId,
    employeeName: opts.employeeName ?? 'Employee',
    clockIn: d,
    clockOut: d,
    jobId: null,
    notes: reason,
    clockInAddress: null,
    clockOutAddress: null,
    clockInLat: null,
    clockInLng: null,
    clockOutLat: null,
    clockOutLng: null,
    isAbsentDay: true,
    isLeaveDay: false,
    statusNote: reason,
    regularHours: 0,
    overtimeHours: 0,
    isLate: false,
    lateMinutes: 0,
    isLeftEarly: false,
    earlyMinutes: 0,
    isOpen: false,
  }
}

export function leaveDaySession(
  dateStr: string,
  leaveType: string,
  opts: PunchSessionOptions,
): PunchSessionRow {
  const d = new Date(`${dateStr}T00:00:00`)
  return {
    employeeId: opts.employeeId,
    employeeName: opts.employeeName ?? 'Employee',
    clockIn: d,
    clockOut: d,
    jobId: null,
    notes: leaveType,
    clockInAddress: null,
    clockOutAddress: null,
    clockInLat: null,
    clockInLng: null,
    clockOutLat: null,
    clockOutLng: null,
    isAbsentDay: false,
    isLeaveDay: true,
    statusNote: leaveType,
    regularHours: 0,
    overtimeHours: 0,
    isLate: false,
    lateMinutes: 0,
    isLeftEarly: false,
    earlyMinutes: 0,
    isOpen: false,
  }
}

/** Merge absence/leave days into sessions for dates without punch sessions. */
export function mergeNonWorkDays(
  sessions: PunchSessionRow[],
  absences: { date: string; reason?: string | null; note?: string | null }[],
  leaveDays: { date: string; leave_type: string }[],
  opts: PunchSessionOptions,
): PunchSessionRow[] {
  const workDates = new Set(
    sessions
      .filter(s => !s.isAbsentDay && !s.isLeaveDay)
      .map(s => toLocalDateStr(s.clockIn)),
  )
  const extra: PunchSessionRow[] = []
  for (const a of absences) {
    if (workDates.has(a.date)) continue
    const note = [a.reason, a.note].filter(Boolean).join(' – ') || 'Absent'
    extra.push(absentDaySession(a.date, note, opts))
    workDates.add(a.date)
  }
  for (const l of leaveDays) {
    if (workDates.has(l.date)) continue
    extra.push(leaveDaySession(l.date, l.leave_type || 'Leave', opts))
    workDates.add(l.date)
  }
  return [...sessions, ...extra].sort((a, b) => b.clockIn.getTime() - a.clockIn.getTime())
}

export function toLocalDateStr(d: Date): string {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

export function fmtSessionTime(d: Date | null, kind: 'in' | 'out' | 'absent' | 'leave'): string {
  if (kind === 'absent') return 'Absent'
  if (kind === 'leave') return 'On Leave'
  if (!d) return '—'
  return d.toLocaleTimeString('en-ZA', { hour: '2-digit', minute: '2-digit', hour12: true })
}

export function fmtSessionDate(d: Date): string {
  return d.toLocaleDateString('en-ZA', {
    weekday: 'short', day: 'numeric', month: 'short', year: 'numeric',
  })
}

export function lateFlag(s: PunchSessionRow): string {
  if (!s.isLate) return ''
  const h = Math.floor(s.lateMinutes / 60)
  const m = s.lateMinutes % 60
  const parts = h > 0 ? (m > 0 ? `${h}h ${m}m` : `${h}h`) : `${m}m`
  return `Late ${parts}`
}

export function earlyFlag(s: PunchSessionRow): string {
  if (!s.isLeftEarly) return ''
  const h = Math.floor(s.earlyMinutes / 60)
  const m = s.earlyMinutes % 60
  const parts = h > 0 ? (m > 0 ? `${h}h ${m}m` : `${h}h`) : `${m}m`
  return `Left Early ${parts}`
}

export function totalHrsDisplay(s: PunchSessionRow): string {
  if (s.isAbsentDay || s.isLeaveDay) return '0.0h'
  return `${(s.regularHours + s.overtimeHours).toFixed(1)}h`
}

export function locationDisplay(address: string | null, lat: number | null, lng: number | null): string {
  if (address) return address
  if (lat != null && lng != null) return `${lat.toFixed(4)}, ${lng.toFixed(4)}`
  return '—'
}
